File: BinaryTree/csharp/BreadthFirstValues.cs
using System.Collections.Generic;

namespace BinaryTreeSolutions;

public static class BreadthFirstValues
{
    public static IList<int> Solve(TreeNode? root)
    {
        var values = new List<int>();
        if (root is null)
        {
            return values;
        }

        var queue = new Queue<TreeNode?>();
        queue.Enqueue(root);
        while (queue.Count > 0)
        {
            var current = queue.Dequeue();
            if (current is null)
            {
                continue;
            }

            values.Add(current.Val);
            if (current.Left is not null) queue.Enqueue(current.Left);
            if (current.Right is not null) queue.Enqueue(current.Right);
        }

        return values;
    }
}

File: BinaryTree/java/breadthFirstValues.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;
import java.util.Queue;

class BreadthFirstValues {
  public static List<Integer> breadthFirstValues(TreeNode root) {
    List<Integer> values = new ArrayList<>();
    if (root == null) {
      return values;
    }

    Queue<TreeNode> queue = new ArrayDeque<>();
    queue.add(root);

    while (!queue.isEmpty()) {
      TreeNode current = queue.remove();
      values.add(current.val);
      if (current.left != null) queue.add(current.left);
      if (current.right != null) queue.add(current.right);
    }

    return values;
  }

  public static void run() {
    // intentionally empty
  }
}
